### DISCUSSION TITLE
2 Node High Availability  cluster for nats stream

### DISCUSSION THREAD

**Discussion by msg2man** [nats-io/nats-server#2869] [General]
I need to create a HA cluster for nats jetstream, my limitations are:-
- only 2 nodes are there, hence using the standard nats streaming cluster with odd number of nodes (to achieve raft quorum) is not possible.
- there is no shared storage between both the nodes or a common nfs mount, so using a fault tolerance nats cluster is also not possible.

So is there a better way of achieving the HA clustering with 2 nodes, or explicitly rsyncing the stream store periodically to the standby node  would be the only way for this ?

Thank you.

**Comment by derekcollison** [comment]
For JetStream and a 2-node system you have data durability but if any of the servers has an issue the system and all streams will be blocked until both servers are back and running.

**Comment by ashupednekar** [comment]
If all you need is two independent jetstream containers running on the two servers, and do not need the raft consistency guarantees, you can replicate all events from one server to the other

Here'a how you can go about it:  https://github.com/ashupednekar/natsriot

There's one flaw right now, where the number of messages grows exponentially due to cyclic duplicate writes

**Reply by ashupednekar** [reply]
@derekcollison any suggestions on this? Ofc this is not the right way to replicate data in nats, it's only for niche cases with this unique scenario

**Reply by neilalexander** [reply]
A better (and supported) replication path would be to use a leafnode connection between the two servers and a sourcing/mirroring configuration in the stream. That way if there is an interruption to one or other server, the sourcing/mirroring would fill the gap automatically. 

https://docs.nats.io/nats-concepts/jetstream/source_and_mirror

**Reply by ashupednekar** [reply]
Say I configure the server 2 as a leaf node to server 1 and create a stream with the same name in server 2 as a mirror stream, all activity from 1 will be replicated to 2, is that it? What about writes to 2?
